client/state/utils/create-reducer.js

```diff
@@ -3,40 +3,24 @@
  */
 import { DESERIALIZE, SERIALIZE } from 'state/action-types';
 import { withoutPersistence } from './without-persistence';
-import { withSchemaValidation } from './schema-utils';
+import { createBaseReducer } from './create-base-reducer';
 
 /**
  * Returns a reducer function with state calculation determined by the result
  * of invoking the handler key corresponding with the dispatched action type,
- * passing both the current state and action object. Defines default
- * serialization (persistence) handlers based on the presence of a schema.
+ * passing both the current state and action object.
  *
  * @param  {*}        initialState   Initial state
  * @param  {Object}   handlers       Object mapping action types to state action handlers
- * @param  {?Object}  schema         JSON schema object for deserialization validation
  * @return {Function}                Reducer function
  */
-export function createReducer( initialState, handlers, schema ) {
-	const reducer = ( state = initialState, action ) => {
-		const { type } = action;
-
-		if ( 'production' !== process.env.NODE_ENV && 'type' in action && ! type ) {
-			throw new TypeError(
-				'Reducer called with undefined type.' +
-					' Verify that the action type is defined in state/action-types.js'
-			);
-		}
-
-		if ( handlers.hasOwnProperty( type ) ) {
-			return handlers[ type ]( state, action );
-		}
-
-		return state;
-	};
-
-	if ( schema ) {
-		return withSchemaValidation( schema, reducer );
+export function createReducer( initialState, handlers, ...args ) {
```

Comment:
Rather than a rest args, what renaming the third param and checking `typeof __deprecatedArg__ !== 'undefined'`?

```suggestion
export function createReducer( initialState, handlers, __deprecatedArg__ ) {
```

Reply:
I believe `createReducer` has had limited value for some time. Does this provide anything helpful?

What if we work to remove `createReducer` completely, instead wrapping "standard" switch-based reducer function with `withSchemaValidation` as necessary?

Reply:
> Rather than a rest args, what renaming the third param and checking `typeof __deprecatedArg__ !== 'undefined'?`

Done! Although I don't think `typeof` is necessary here, so I simplified things.

> I believe `createReducer` has had limited value for some time. Does this provide anything helpful?
>
> What if we work to remove `createReducer` completely, instead wrapping "standard" switch-based reducer function with `withSchemaValidation` as necessary?

It looks to me like `createReducer` still provides some benefits around sanity checking and serialisation, although my experience here is very limited. I'm happy to make the change here or in a follow-up PR if folks agree!

Reply:
I agree that `createReducer` is not terribly useful. Simple `switch` statement provides almost the same experience and is likely to be faster, too, as there are no lambda functions and the code is very close to C-like style.

@sgomes Does the `createReducerWithValidation` provide anything valuable over composing two simpler functions to achieve the result?
```js
const plainReducer = ( state, action ) => {
  switch ( action.type ) { ... }
};
const validatedReducer = withSchemaValidation( schema )( plainReducer );
const combinedReducer = combineReducers( { foo: validatedReducer } );
```

Reply:
`createBaseReducer`, which I extracted out and is used by both `createReducer` and `createReducerWithValidation` only seems to add a sanity check for actions with invalid (falsy) types. As for `createReducerWithValidation`, it only checks to ensure that a schema is provided.

It sounds like we have a consensus on `createReducer` and `createReducerWithValidation` having outlived their usefulness 🧟‍♂️

Should I tackle that in this PR, or a separate one?

Reply:
> Should I tackle that in this PR, or a separate one, @jsnajdr and @sirreal?

The work here, by itself, seems valuable. I'd also suspect this is incremental progress towards removing `createReducer` and friends. I'd vote to merge this, then continue with follow-ups.

Reply:
I agree, let's merge and do incremental improvements later. 100+ changed files already makes a nice PR 😄

Reply:
Sounds good, thank you both! 👍

Reply:
@jsnajdr @sirreal I started looking into this more closely today, and it looks like `createReducer` adds some persistence code. Namely, if there are no specific handlers for serialization/deserialization, it wraps the reducer with the default ones:

```js
...
if ( ! handlers[ SERIALIZE ] && ! handlers[ DESERIALIZE ] ) {
  return withoutPersistence( reducer );
}
...
```

I'm not hugely familiar with this part of the codebase, so what would the consequences be of not doing so? And does this mean that we'd need to wrap every plain reducer with `withoutPersistence`, if we're trying to get rid of `createReducer`, e.g.?

```js
const isFetchingSettings = ( state = false, action ) => {
  switch ( action.type ) {
    case ACCOUNT_RECOVERY_SETTINGS_FETCH:
      return true;
    case ACCOUNT_RECOVERY_SETTINGS_FETCH_SUCCESS:
      return false;
    case ACCOUNT_RECOVERY_SETTINGS_FETCH_FAILED:
      return false;
  }
};

export default combineReducers( {
  settings,
  withoutPersistence( isFetchingSettings ),
} );
```

Reply:
What the `withoutPersistence` wrapper does is that it makes the `SERIALIZE` action dispatch return `undefined`, meaning "we don't want to persist this part of the tree". Without it, the reducer's state will be saved to IndexedDB.

Fortunately, the `combineReducer` function automatically wraps every reducer with `withoutPersistence` before making it part of the combination. That happens only when the reducer function doesn't have a `hasCustomPersistence = true` property. That means that persistence for that reducer is already handled (i.e., one of the persistence wrapper has been applied to it) and `combineReducer` can combine it as it is.

Therefore, wrapping all reducers with `withoutPersistence` is not necessary.

Reply:
That's perfect, thank you for the thorough explanation, @jsnajdr! I can forward now, then! 👍